contracts/finance/andromeda-mint-burn/src/execute.rs

```diff
@@ -0,0 +1,456 @@
+        ctx.deps.branch(),
+        ctx.env.clone(),
+        SEND_NFT_ACTION,
+        ctx.info.sender.clone(),
+    )?;
+
+    let received_token_id = receive_msg.token_id;
+    let original_sender = AndrAddr::from_string(receive_msg.sender);
+
+    match from_json(&receive_msg.msg)? {
+        Cw721HookMsg::FillOrder { order_id } => execute_fill_order(
+            ctx,
+            order_id,
+            original_sender,
+            Uint128::one(),
+            Some(received_token_id),
+        ),
+    }
+}
+
+pub fn handle_receive_cw20(
+    mut ctx: ExecuteContext,
+    receive_msg: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    ADOContract::default().is_permissioned(
+        ctx.deps.branch(),
+        ctx.env.clone(),
+        SEND_CW20_ACTION,
+        ctx.info.sender.clone(),
+    )?;
+
+    let original_sender = AndrAddr::from_string(receive_msg.sender);
+    let cw20_amount = receive_msg.amount;
+
+    match from_json(&receive_msg.msg)? {
+        Cw20HookMsg::FillOrder { order_id } => {
+            execute_fill_order(ctx, order_id, original_sender, cw20_amount, None)
+        }
+    }
+}
+
+pub fn execute_create_order(
+    ctx: ExecuteContext,
+    requirements: Vec<ResourceRequirement>,
+    output: Resource,
+) -> Result<Response, ContractError> {
+    let ExecuteContext {
+        mut deps,
+        info,
+        env,
+        ..
+    } = ctx;
+    let sender = info.sender.clone();
+
+    let order_id = NEXT_ORDER_ID.load(deps.storage)?;
+
+    if ORDERS
+        .may_load(deps.storage, order_id.clone().u128())?
+        .is_some()
+    {
+        return Err(ContractError::CustomError {
+            msg: "Order ID already exists".to_string(),
+        });
+    }
+
+    // Initialize all resource requirements with empty deposit tracking
+    let mut initialized_requirements = Vec::new();
+    for reqirement in requirements {
+        reqirement.validate()?;
+        match reqirement.clone().resource {
+            Resource::Cw20Token { cw20_addr } => {
+                ADOContract::default().is_permissioned(
+                    deps.branch(),
+                    env.clone(),
+                    SEND_CW20_ACTION,
+                    cw20_addr,
+                )?;
+            }
+            Resource::Nft { cw721_addr, .. } => {
+                ADOContract::default().is_permissioned(
+                    deps.branch(),
+                    env.clone(),
+                    SEND_NFT_ACTION,
+                    cw721_addr,
+                )?;
+            }
+        }
+
+        let new_req = ResourceRequirement {
+            resource: reqirement.clone().resource,
+            amount: reqirement.amount,
+            deposits: HashMap::new(), // No deposits yet
+        };
+        initialized_requirements.push(new_req);
+    }
+
+    let new_order = OrderInfo {
+        requirements: initialized_requirements,
+        output,
+        order_status: OrderStatus::NotCompleted,
+        output_recipient: None,
```

Comment:
Currently there's no way of having this be other than the original sender. We could have an optional output_recipient in the function's arguments.